src/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
    */
   public static String validateCanal(final Territory territory, final Route route, final Collection<Unit> units,
       final PlayerID player, final GameData data) {
+    String failureMessage = null;
     final Set<CanalAttachment> canalAttachments = CanalAttachment.get(territory);
-    if (canalAttachments.isEmpty()) {
-      return null;
-    }
-    final Iterator<CanalAttachment> iter = canalAttachments.iterator();
-    while (iter.hasNext()) {
-      final CanalAttachment attachment = iter.next();
-      if (attachment == null) {
+    for (final CanalAttachment canalAttachment : canalAttachments) {
+      if (!needToCheckCanal(canalAttachment, route, data)) {
         continue;
       }
-      if (route != null) {
-        boolean mustCheck = false;
-        Territory last = null;
-        final Set<Territory> connectionToCheck = CanalAttachment.getAllCanalSeaZones(attachment.getCanalName(), data);
-        for (final Territory current : route.getAllTerritories()) {
-          if (last != null) {
-            final Collection<Territory> lastTwo = new ArrayList<Territory>();
-            lastTwo.add(last);
-            lastTwo.add(current);
-            mustCheck = lastTwo.containsAll(connectionToCheck);
-            if (mustCheck) {
-              break;
-            }
-          }
-          last = current;
-        }
-        if (!mustCheck) {
-          continue;
+      failureMessage = canPassThroughCanal(canalAttachment, units, player, data);
+      if (failureMessage == null) {
+        return null;
+      }
+    }
+    return failureMessage;
+  }
+
+  /**
```

Comment:
Agree. I'll change them to non-java doc though doesn't make a huge difference since no one really reads are java docs and a good portion of even public methods have no java doc.